File: week02/박재은/BOJ_17182.py
#우주 탐사선
def DFS(cnt, start, total):
    global ans
    if cnt == n:
        ans = min(ans, total)
        return
    for i in range(n):
        if not visited[i]:
            visited[i] = True
            DFS(cnt+1, i, total + time[start][i])
            visited[i] = False

n, m = map(int, input().split())
time = [list(map(int, input().split())) for _ in range(n)]
for k in range(n):
    for i in range(n):
        for j in range(n):
            time[i][j] = min(time[i][j], time[i][k] + time[k][j])

visited = [False] * n
visited[m] = True
ans = 2**31 - 1
DFS(1, m, 0)
print(ans)

File: week02/정승호/BOJ_17182.java
package 플로이드;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ_17182 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken()); // 행성의 개수
        int start = Integer.parseInt(st.nextToken()); // 시작 행성

        int[][] dist = new int[n][n];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                dist[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        // 플로이드 워셜 알고리즘으로 모든 정점 간의 최단 거리 구하기
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (dist[i][j] > dist[i][k] + dist[k][j]) {
                        dist[i][j] = dist[i][k] + dist[k][j];
                    }
                }
            }
        }
        boolean[] visited = new boolean[n];
        visited[start] = true;
        int ans = dfs(start, visited, dist, n, 1);
        System.out.println(ans);
    }

    static int dfs(int cur, boolean[] visited, int[][] dist, int n, int cnt) {
        if (cnt == n) {
            return 0; // 모든 행성을 방문했으면 비용 0 반환
        }

        int result = Integer.MAX_VALUE;
        for (int next = 0; next < n; next++) {
            if (!visited[next]) {
                visited[next] = true;
                int cost = dist[cur][next] + dfs(next, visited, dist, n, cnt + 1);
                result = Math.min(result, cost);
                visited[next] = false;
            }
        }
        return result;
    }
}
